do_print: Use new utf8_to_bytes_temp_pv()

khwilliamson opened the pull request:
* This set of changes does not require a perldelta entry.

tonycoz commented on doio.c:
```diff
@@ -2215,35 +2215,23 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
         STRLEN len;
         /* Do this first to trigger any overloading.  */
         const char *tmps = SvPV_const(sv, len);
-        U8 *tmpbuf = NULL;
```
commit message: "Use new **utfj8**_to_bytes_temp_pv"

tonycoz commented on doio.c:
```diff
@@ -2215,35 +2215,23 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
-                       utf8 */
-                assert((char *)result == tmps);
+            if (! utf8_to_bytes_temp_pv((const U8**) &tmps, &len)) {
+                /* Non-utf8 output stream, but string only representable in
+                   utf8 */
```
This has a similar problem to do_chomp() - print may be called with a large list, which could leave a large number of temporary allocations active beyond their use until the next LEAVE, and significantly expand the save stack.

khwilliamson committed:
doio.c Change a variable from char* to U8*

This removes 6 casts and adds 1, resulting in cleaner looking code.

tonycoz commented on embed.fnc:
```diff
@@ -3716,7 +3716,7 @@ Cp	|bool	|utf8_to_bytes_ |NN U8 **s_ptr				\
 				|NN U8 const **s_ptr			\
 				|NN STRLEN *lenp			\
-				|NN U8 *free_me
+				|NN const U8 *free_me
```
I may have missed this, since it's purely a documentation issue, but `free_me` isn't a pointer to U8, it certainly isn't a pointer to const U8.

`free_me` would ideally be declared as `void **` but the existing users are passing in a `U8 **` which makes that more reasonable here.

And it shouldn't be const in any way, free() requires a non-const pointer, it's only a bug that Safefree() casts const away.  Your modification to the macro below also casts const away.

Using inline functions instead of macros here would have made the type mismatches a lot more obvious.

tonycoz commented on doio.c:
```diff
@@ -2214,36 +2214,37 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
         STRLEN len;
         /* Do this first to trigger any overloading.  */
-        const char *tmps = SvPV_const(sv, len);
+        const U8 *tmps = (U8 *) SvPV_const(sv, len);
```
For consistency the cast should be to `(const U8 *)`, since the rest of  SvPV_const() is a const pointer, you're casting const away just to add it back in. (makes me wish for C++ const_cast, static_cast etc)

tonycoz commented on doio.c:
```diff
@@ -2259,7 +2260,7 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
          * at which we would get EPERM.  Note that when using buffered
          * io the write failure can be delayed until the flush/close. --jhi */
-        if (len && (PerlIO_write(fp,tmps,len) == 0))
+        if (len && (PerlIO_write(fp,(char *) tmps,len) == 0))
```
PerlIO_write() takes a `const void *` for the buffer, you don't need the cast.

tonycoz replied:
This is still the case, no cast is required

khwilliamson replied:
sorry; I somehow missed seeing your first comment

tonycoz commented on doio.c:
```diff
@@ -2215,36 +2215,29 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
+         * allocated memory that tmps will also be changed to point to, so
+         * Safefree(free_me) will free it.  This saves having to have extra
+         * logic. */
+        const U8 *free_me = NULL;
```
Per my comments on the other commits, this should not be const.

khwilliamson committed:
do_print(): Use bytes_to_utf8_free_me, utf8_to_bytes_new_pv

These new functions do the work this used to duplicate, except they
allocate memory only if necessary, instead of always, as previously
